sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/CryptographyAsyncClient.java

```diff
@@ -943,36 +936,57 @@ public Mono<VerifyResult> verifyData(SignatureAlgorithm algorithm, byte[] data,
+            Mono<JsonWebKey> jsonWebKeyMono = CryptographyUtils.SECRETS_COLLECTION.equals(implClient.getKeyCollection())
+                ? implClient.getSecretKeyAsync() : getKey().map(KeyVaultKey::getKey);
+
+            return jsonWebKeyMono
+                .map(this::validateJsonWebKeyAndCreateLocalClient)
+                .onErrorResume(HttpResponseException.class, this::handleKeyRetrievalError);
+        } else {
+            return Mono.defer(() -> Mono.just(true));
         }
+    }
+
+    private boolean validateJsonWebKeyAndCreateLocalClient(JsonWebKey jsonWebKey) {
+        key = jsonWebKey;
+
+        if (jsonWebKey.isValid()) {
+            // Create a local key cryptography client if one has not been created yet.
+            if (localKeyCryptographyClient == null) {
+                try {
+                    localKeyCryptographyClient = initializeCryptoClient(jsonWebKey, implClient, LOGGER);
+                } catch (RuntimeException e) {
+                    localOperationNotSupported = true;
+
+                    LOGGER.warning("Defaulting to service use for cryptographic operations.", e);
 
-        if (key == null && implClient.getKeyCollection() != null) {
-            Mono<JsonWebKey> jsonWebKeyMono = CryptographyUtils.SECRETS_COLLECTION.equals(implClient.getKeyCollection())
-                ? getSecretKey()
-                : getKey().map(KeyVaultKey::getKey);
-
-            return jsonWebKeyMono.map(jsonWebKey -> {
-                key = jsonWebKey;
-
-                if (jsonWebKey.isValid()) {
-                    if (localKeyCryptographyClient == null) {
-                        try {
-                            localKeyCryptographyClient = initializeCryptoClient(jsonWebKey, implClient, LOGGER);
-                        } catch (RuntimeException e) {
-                            localOperationNotSupported = true;
-                            LOGGER.warning("Defaulting to service use for cryptographic operations.", e);
-
-                            return false;
-                        }
-                    }
-
-                    return true;
-                } else {
                     return false;
                 }
-            });
+            }
+
+            // Valid key is locally available and local client was created.
+            return true;
         } else {
-            return Mono.just(true);
+            return false;
+        }
+    }
+
+    private Mono<Boolean> handleKeyRetrievalError(HttpResponseException e) {
+        if (e.getResponse().getStatusCode() == 403) {
+            localOperationNotSupported = true;
+
+            LOGGER.warning("Could not retrieve key from service to perform cryptographic operations locally. "
+                + "Will use service-side cryptography instead.", e);
+
+            return Mono.defer(() -> Mono.just(false));
+        } else {
+            LOGGER.error("Could not retrieve key from service.");
+
+            throw LOGGER.logExceptionAsError(e);
```

Comment:
Why do we throw here instead of either using `Mono.error` or returning `Mono.just(false)` as we do in other spots.

Reply:
Hmmm... that's true, if we delegate the operation to the service no matter why we were not able to retrieve the key we should let the service tell users if the operation can or cannot be performed on their end.